[BUG] Intermittent test failures

### Describe your issue

There are two intermittent test failures on the CI that it would be nice not to have

- [x] `bytearray_coercion` on PyPy. This looks like `PyBytearray_AS_BYTES` either isn't returning a `\0` terminated string, or that the validity of the `char*` it returns isn't being maintained for long enough. We hold a real reference rather than a borrowed reference to the argument so its probably not that. I couldn't reproduce it locally (running about 10 times) with pypy3.10-v7.3.12-linux64, which was picked because it's the most up-to-date version already on my disk.

    Example of failing output

    ```
    File "/home/runner/work/cython/cython/TEST_TMP/5/run/c/bytearray_coercion/bytearray_coercion.pypy39-pp73-x86_64-linux-gnu.so", line ?, in bytearray_coercion.coerce_to_charptr
    Failed example:
        coerced == b or (coerced, b)
    Expected:
        True
    Got:
        (bytearray(b'abc\x1b\xf1\x7f'), bytearray(b'abc'))
   ```

    @mattip - any thoughts on this one
- [ ] Line profiling on Windows + Python 3.12. I think we're suspicious all forms of profiling on Python 3.12, but line profiling is possibly using the old mechanism to OK. It only looks to fail occasionally on Windows. Example of failing output:

    ```
     ======================================================================
    FAIL: disable_trace (line_trace)
    Doctest: line_trace.disable_trace
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "C:\hostedtoolcache\windows\Python\3.12.0\x64\Lib\doctest.py", line 2228, in runTest
        raise self.failureException(self.format_failure(new.getvalue()))
    AssertionError: Failed doctest test for line_trace.disable_trace
      File "D:\a\cython\cython\TEST_TMP\1\run\c\line_trace\line_trace.cp312-win_amd64.pyd", line unknown line number, in disable_trace

    ----------------------------------------------------------------------
    File "D:\a\cython\cython\TEST_TMP\1\run\c\line_trace\line_trace.cp312-win_amd64.pyd", line ?, in line_trace.disable_trace
    Failed example:
        disable_trace(py_add, 1, 2, with_sys=True)
    Expected:
        [('call', 0), ('line', 1)]
    Got:
        [('call', 0), ('call', 0)]
    ```
    Output seems reproducible (when it fails). I suspect we're OK  to just disable this test.

--- a/tests/run/line_trace.pyx
+++ b/tests/run/line_trace.pyx
@@ -4,6 +4,8 @@
 # tag: trace
 
 import sys
+import gc
+from contextlib import contextmanager
 
 from cpython.ref cimport PyObject, Py_INCREF, Py_XDECREF
 
@@ -180,6 +182,17 @@ def py_return(retval=123): return retval
 """, plain_python_functions)
 
 
+@contextmanager
+def gc_off():
+    was_enabled = gc.isenabled()
+    gc.disable()
+    try:
+        yield
+    finally:
+        if was_enabled:
+            gc.enable()
+
+
 def run_trace(func, *args, bint with_sys=False):
     """
     >>> py_add = plain_python_functions['py_add']
@@ -229,17 +242,18 @@ def run_trace(func, *args, bint with_sys=False):
     """
     trace = []
     trace_func = _create_trace_func(trace)
-    if with_sys:
-        sys.settrace(trace_func)
-    else:
-        PyEval_SetTrace(<Py_tracefunc>trace_trampoline, <PyObject*>trace_func)
-    try:
-        func(*args)
-    finally:
+    with gc_off():
         if with_sys:
-            sys.settrace(None)
+            sys.settrace(trace_func)
         else:
-            PyEval_SetTrace(NULL, NULL)
+            PyEval_SetTrace(<Py_tracefunc>trace_trampoline, <PyObject*>trace_func)
+        try:
+            func(*args)
+        finally:
+            if with_sys:
+                sys.settrace(None)
+            else:
+                PyEval_SetTrace(NULL, NULL)
     return trace
 
 
@@ -278,24 +292,25 @@ def run_trace_with_exception(func, bint with_sys=False, bint fail=False):
     """
     trace = ['cy_try_except' if fail else 'NO ERROR']
     trace_func = _create__failing_line_trace_func(trace) if fail else _create_trace_func(trace)
-    if with_sys:
-        sys.settrace(trace_func)
-    else:
-        PyEval_SetTrace(<Py_tracefunc>trace_trampoline, <PyObject*>trace_func)
-    try:
-        try:
-            retval = cy_try_except(func)
-        except ValueError as exc:
-            print("%s(%r)" % (type(exc).__name__, str(exc)))
-        except AttributeError as exc:
-            print("%s(%r)" % (type(exc).__name__, str(exc)))
-        else:
-            print('OK: %r' % retval)
-    finally:
+    with gc_off():
         if with_sys:
-            sys.settrace(None)
+            sys.settrace(trace_func)
         else:
-            PyEval_SetTrace(NULL, NULL)
+            PyEval_SetTrace(<Py_tracefunc>trace_trampoline, <PyObject*>trace_func)
+        try:
+            try:
+                retval = cy_try_except(func)
+            except ValueError as exc:
+                print("%s(%r)" % (type(exc).__name__, str(exc)))
+            except AttributeError as exc:
+                print("%s(%r)" % (type(exc).__name__, str(exc)))
+            else:
+                print('OK: %r' % retval)
+        finally:
+            if with_sys:
+                sys.settrace(None)
+            else:
+                PyEval_SetTrace(NULL, NULL)
     return trace[1:]
 
 
@@ -312,11 +327,12 @@ def fail_on_call_trace(func, *args):
     """
     trace = []
     trace_func = _create_failing_call_trace_func(trace)
-    PyEval_SetTrace(<Py_tracefunc>trace_trampoline, <PyObject*>trace_func)
-    try:
-        func(*args)
-    finally:
-        PyEval_SetTrace(NULL, NULL)
+    with gc_off():
+        PyEval_SetTrace(<Py_tracefunc>trace_trampoline, <PyObject*>trace_func)
+        try:
+            func(*args)
+        finally:
+            PyEval_SetTrace(NULL, NULL)
     assert not trace
 
 
@@ -376,20 +392,21 @@ def fail_on_line_trace(fail_func, add_func, nogil_add_func):
     trace = ['NO ERROR']
     exception = None
     trace_func = _create__failing_line_trace_func(trace)
-    PyEval_SetTrace(<Py_tracefunc>trace_trampoline, <PyObject*>trace_func)
-    try:
-        x += 1
-        add_func(1, 2)
-        x += 1
-        if fail_func:
-            trace[0] = fail_func  # trigger error on first line
-        x += 1
-        nogil_add_func(3, 4)
-        x += 1
-    except Exception as exc:
-        exception = str(exc)
-    finally:
-        PyEval_SetTrace(NULL, NULL)
+    with gc_off():
+        PyEval_SetTrace(<Py_tracefunc>trace_trampoline, <PyObject*>trace_func)
+        try:
+            x += 1
+            add_func(1, 2)
+            x += 1
+            if fail_func:
+                trace[0] = fail_func  # trigger error on first line
+            x += 1
+            nogil_add_func(3, 4)
+            x += 1
+        except Exception as exc:
+            exception = str(exc)
+        finally:
+            PyEval_SetTrace(NULL, NULL)
     if exception:
         print(exception)
     else:
@@ -397,15 +414,6 @@ def fail_on_line_trace(fail_func, add_func, nogil_add_func):
     return trace
 
 
-def skip_on_win_py3_12_0(func):
-    if sys.version_info[:3] == (3, 12, 0) and sys.platform == "win32":
-        # This test is mysteriously failing on the CI only. Disable
-        # it for now and hope that the next minor release fixes it.
-        return None
-    return func
-
-
-@skip_on_win_py3_12_0
 def disable_trace(func, *args, bint with_sys=False):
     """
     >>> py_add = plain_python_functions["py_add"]
@@ -426,15 +434,16 @@ def disable_trace(func, *args, bint with_sys=False):
     """
     trace = []
     trace_func = _create_disable_tracing(trace)
-    if with_sys:
-        sys.settrace(trace_func)
-    else:
-        PyEval_SetTrace(<Py_tracefunc>trace_trampoline, <PyObject*>trace_func)
-    try:
-        func(*args)
-    finally:
+    with gc_off():
         if with_sys:
-            sys.settrace(None)
+            sys.settrace(trace_func)
         else:
-            PyEval_SetTrace(NULL, NULL)
+            PyEval_SetTrace(<Py_tracefunc>trace_trampoline, <PyObject*>trace_func)
+        try:
+            func(*args)
+        finally:
+            if with_sys:
+                sys.settrace(None)
+            else:
+                PyEval_SetTrace(NULL, NULL)
     return trace

